src/core/net/SRP_CODER.md

```diff
@@ -0,0 +1,349 @@
+# SRP Coder
+
+This document specifies SRP coder used to encode an SRP message into a compact, compressed format, reducing the message size. On the server, the received coded message can be decoded to reconstruct an exact replica of the original message.
+
+The SRP protocol uses DNS resource record format to convey information about registered services and host addresses. The use of these records and adherence to the DNS Update general format introduces overhead and can lead to larger SRP message sizes, which is undesirable for Thread network. The SRP coder aims to optimize the message, resulting in much smaller messages containing all the same information reducing overall protocol overhead.
+
+## Compact Integer Format
+
+This format efficiently encodes unsigned integers of varying bit-lengths (e.g., `uint16`, `uint32`). The unsigned integer value is encoded as one or more segments:
+
+- The first segment can be 2-bit to 8-bit long. Its bit-length is specified based on the context where the compact encoding is used. If not explicitly specified, an 8-bit length MUST be assumed for the first segment.
+- Subsequent segments are one byte (8-bit) long.
+
+The most significant bit in each segment (regardless of its bit-length) is the "continuation bit":
+
+- A value of `1` indicates more segments follow.
+- A value of `0` indicates that this is the last segment.
+
+The remaining bits after the most significant bit provide the numerical `uint` bit value chunks and follow big-endian order.
+
+```
+    First segment (2-8 bits)          Next segments (8 bits)
+   +---+---+---+---+---+---+         +---+---+---+---+---+---+---+---+
```

Comment:
I think part of what's confusing me here is that in fact every segment is 8 bits. It's just that the first segment may contain one or more zero bits to the left of the most significant one bit of the original integer. So I would suggest diagramming the first segment as 8 bits and then explicitly mentioning the padding.

Reply:
> I think part of what's confusing me here is that in fact every segment is 8 bits.

That's not quite the case. The first segment can be restricted to fewer bits, allowing for the use of space within a previous field. For example, the remaining bits in a "dispatch byte" can be utilized as the first segment of a compact integer that follows.

The "label dispatch" with type `0b10 <6-bit offset first seg>` (which references a previously encoded label within the message) uses this approach, with the lower 6 bits serving as the first segment to encode the offset (to the previous label).